File: algorithms/Permutation Sequence/Permutation Sequence.js
const dfs = (path, visited, result, numbers, limit) => {
    // return if we already reached the permutation needed
    if(result.length === limit) {
        return;
    }
    
    // commit the result
    if(path.length === numbers.length) {
        result.push(path.join(''))
        return;
    }
    
    // easier to reason and less prone to miss the -1 offset of normal for loop
    for(const [index, number] of numbers.entries()) {
        if(visited[index]) continue;
        
        path.push(number);
        visited[index] = true;
        dfs(path, visited, result, numbers);
        path.pop();
        visited[index] = false;
    }
}

var getPermutation = function(n, k) {
    const numbers = Array.from({length: n}, (_, i) => i + 1);
    let visitedNumbers = Array.from(numbers, () => false);
    let result = [];
    dfs([], visitedNumbers, result, numbers, k);
    return result[k - 1];
};

File: scripts/algorithms/P/Permutation Sequence/Permutation Sequence.py
// Runtime: 38 ms (Top 71.08%) | Memory: 17.30 MB (Top 15.2%)

class Solution:
    def getPermutation(self, n: int, k: int) -> str:
        nums = [i for i in range(1, n+1)] # list of numbers from 1 to n
        factorial = [1] * n
        for i in range(1, n):
            factorial[i] = factorial[i-1] * i
        
        k -= 1
        result = []
        for i in range(n-1, -1, -1):
            index = k // factorial[i]
            result.append(str(nums[index]))
            nums.pop(index)
            k = k % factorial[i]
        
        return ''.join(result)
